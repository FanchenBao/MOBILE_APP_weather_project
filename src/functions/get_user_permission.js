//@flow
import {PermissionsAndroid} from 'react-native';

/**
 * Ask for user to grant specific permission.
 * @param {string} permission The specific permission to be granted or denied.
 * @param {string} title Title for the rationale.
 * @param {string} msg Message in the rationale.
 */
async function getUserPermission(
  permission: string,
  title: string,
  msg: string,
) {
  try {
    let userDecision = await PermissionsAndroid.request(permission, {
      // rationale for obtaining fine location permission
      title: title,
      message: msg,
      buttonNeutral: 'Give Permission',
    });
    return userDecision;
  } catch (error) {
    console.error(error);
  }
}

/**
 * Ask for user permission to access device's fine location.
 * @returns A promise that evaluates to user's decision.
 */
const getFineLocationPermission = () =>
  getUserPermission(
    PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
    'Weather Project App Geolocation Permission',
    'Weathr Project App needs access to your geolocation to provide weather information for your current location.',
  );

/**
 * Ask for user permission to read device's external storage.
 * @returns A promise that evaluates to user's decision.
 */
const getReadExternalStoragePermission = () =>
  getUserPermission(
    PermissionsAndroid.PERMISSIONS.READ_EXTERNAL_STORAGE,
    'Weather Project App Read External Storage Permission',
    'Weather Project App needs access to your external storage so that you can set background to your own photo.',
  );

export {
  getUserPermission,
  getFineLocationPermission,
  getReadExternalStoragePermission,
};
