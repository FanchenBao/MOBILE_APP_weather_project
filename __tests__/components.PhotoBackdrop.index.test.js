import {PhotoBackdrop} from '../src/components/PhotoBackdrop/index.js';
import {getReadExternalStoragePermission} from '../src/functions/get_user_permission.js';
import CameraRoll from '@react-native-community/cameraroll';
import React from 'react';
import renderer from 'react-test-renderer';

let mockConsoleLog;
let mockConsoleError;
let mockSetState;
let component;
// set up mock console functions before each tests.
beforeEach(async () => {
  mockConsoleLog = jest.spyOn(global.console, 'log').mockImplementation();
  mockConsoleError = jest.spyOn(global.console, 'error').mockImplementation();

  // use prototype to access setState()
  mockSetState = jest.spyOn(PhotoBackdrop.prototype, 'setState');

  // create a PhotoBackdrop component and use wait() to wait for the component
  // to update its state based on the mock return of
  // getReadExternalStoragePermission.
  component = renderer.create(<PhotoBackdrop />);
  await wait(); // wait for state update
});

// clean up console function mock after each test.
afterEach(() => {
  mockConsoleLog.mockRestore();
  mockConsoleError.mockRestore();
  mockSetState.mockRestore();
});

// mock app's own module
jest.mock('../src/functions/get_user_permission.js');
getReadExternalStoragePermission
  .mockResolvedValueOnce('granted') // test 1
  .mockResolvedValueOnce('custom_denied') // test 2
  .mockRejectedValueOnce('oops'); // test 3

// Mock CameraRoll.getPhotos. This is possible because the interface of this
// library is already mocked in
// @react-native-community/cameraroll/js/__mock__/nativeInterface.js
CameraRoll.getPhotos = jest.fn(prop =>
  Promise.resolve({edges: [{node: {image: {uri: 'some uri'}}}]}),
);

const wait = async () => 'foo'; // a dummy function to simulate waiting

// test 1
test('render PhotoBackdrop with user photo', () => {
  expect(component.toJSON()).toMatchSnapshot(); // uri should be 'some uri'
  expect(mockConsoleLog).not.toHaveBeenCalled();
  expect(mockConsoleError).not.toHaveBeenCalled();
  expect(mockSetState).toHaveBeenCalledTimes(1);
  expect(mockSetState).toHaveBeenLastCalledWith({photoURIs: ['some uri']});
});

// test 2
test('cannot render PhotoBacdrop due to user denied permission', () => {
  expect(component.toJSON()).toMatchSnapshot();
  expect(mockConsoleLog).toHaveBeenCalledTimes(1);
  expect(mockConsoleLog).toHaveBeenCalledWith(
    'Read external storage denied. User decision: custom_denied',
  );
  expect(mockConsoleError).not.toHaveBeenCalled();
  expect(mockSetState).not.toHaveBeenCalled();
});

// test 3
test('Error occurs when getting user permission or photos', () => {
  expect(mockConsoleLog).not.toHaveBeenCalled();
  expect(mockConsoleError).toHaveBeenCalledTimes(1);
  expect(mockConsoleError).toHaveBeenLastCalledWith('oops');
  expect(mockSetState).not.toHaveBeenCalled();
});
