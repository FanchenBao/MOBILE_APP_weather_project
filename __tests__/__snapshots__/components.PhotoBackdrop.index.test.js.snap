// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cannot render PhotoBacdrop due to user denied permission 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "height": 1334,
      "width": 750,
    }
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "dummy/uri",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": 1334,
          "width": 750,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`render PhotoBackdrop with user photo 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "height": 1334,
      "width": 750,
    }
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "some uri",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": 1334,
          "width": 750,
        },
        undefined,
      ]
    }
  />
</View>
`;
